dot/core/process.go

```diff
@@ -0,0 +1,281 @@
+			if err == nil && retData != nil {
+				blockData.Body = retData.AsOptional()
+			}
+		}
+		// receipt
+		if (blockRequest.RequestedData&4)>>2 == 1 {
+			retData, err := s.blockState.GetReceipt(hash)
+			if err == nil && retData != nil {
+				blockData.Receipt = optional.NewBytes(true, retData)
+			}
+		}
+		// message queue
+		if (blockRequest.RequestedData&8)>>3 == 1 {
+			retData, err := s.blockState.GetMessageQueue(hash)
+			if err == nil && retData != nil {
+				blockData.MessageQueue = optional.NewBytes(true, retData)
+			}
+		}
+		// justification
+		if (blockRequest.RequestedData&16)>>4 == 1 {
+			retData, err := s.blockState.GetJustification(hash)
+			if err == nil && retData != nil {
+				blockData.Justification = optional.NewBytes(true, retData)
+			}
+		}
+
+		responseData = append(responseData, blockData)
+	}
+
+	return &network.BlockResponseMessage{
+		ID:        blockRequest.ID,
+		BlockData: responseData,
+	}, nil
+}
+
+// BlockResponseMessage 2
+
+// ProcessBlockResponseMessage attempts to validate and add the block to the
+// chain by calling `core_execute_block`. Valid blocks are stored in the block
+// database to become part of the canonical chain.
+func (s *Service) ProcessBlockResponseMessage(msg *network.BlockResponseMessage) error {
+	log.Debug("[core] received BlockResponseMessage")
+
+	// loop through block data in block response message
+	for _, bd := range msg.BlockData {
+		if bd.Header.Exists() {
+
+			header, err := types.NewHeader(
+				bd.Header.Value().ParentHash,
+				bd.Header.Value().Number,
+				bd.Header.Value().StateRoot,
+				bd.Header.Value().ExtrinsicsRoot,
+				bd.Header.Value().Digest,
+			)
+			if err != nil {
+				return err
+			}
+
+			// if authority, check if first block and set epoch data for next epoch
+			if s.isAuthority {
+				err = s.handleConsensusDigest(header)
+				if err != nil {
+					log.Trace("[core] block is not first block in epoch", "error", err)
+				}
+			}
+		}
+	}
+
+	// send block response message to syncer
+	s.respOut <- msg
+
+	// TODO: should we wait for syncer to handle message before we check for runtime changes...?
+
+	// check if we need to update the runtime
+	err := s.checkForRuntimeChanges()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// BlockAnnounceMessage 3
+
+// ProcessBlockAnnounceMessage creates a block request message from the block
+// announce messages (block announce messages include the header but the full
+// block is required to execute `core_execute_block`).
+func (s *Service) ProcessBlockAnnounceMessage(msg *network.BlockAnnounceMessage) error {
+	log.Debug("[core] received BlockAnnounceMessage")
+
+	// create header from message
+	header, err := types.NewHeader(
+		msg.ParentHash,
+		msg.Number,
+		msg.StateRoot,
+		msg.ExtrinsicsRoot,
+		msg.Digest,
+	)
+	if err != nil {
+		return err
+	}
+
+	// if authority, check if first block and set epoch data for next epoch
+	if s.isAuthority {
```

Comment:
same here

Reply:
it is inside `handleConsensusDigest` because you request that this be consolidated